### PR TITLE
Implementing Background Process Manager

### PR DESCRIPTION
**Pull request by abbykrish**
We created a package called subprocess, under the shared directory for process management to address the issue: #5826 

Some new features 
* You can start a process and configure it in Go using 'NewProcess(..)' 
* You can manage this process using functions like `Pid(), Stop(), Start(), Restart(), Reload(), Save(path string), Signal(int64), Wait() (int64, error)` 
* You can now serialize and deserialize configurations for a process by calling 'Save()' and 'ImportProcess(..)'

### CODE REVIEW

**Comment by stgraber** [comment]
`go fmt ./...` has a few fixes for `bgpm_test.go`.

You also need to add documentation comments so `golint` is happy:
```
stgraber@castiana:~/data/code/lxc/lxd (bgpm)$ golint shared/subprocess/
shared/subprocess/manager.go:9:1: exported function NewProcess should have comment or be unexported
shared/subprocess/manager.go:20:1: exported function ImportProcess should have comment or be unexported
shared/subprocess/proc.go:13:6: exported type Process should have comment or be unexported
shared/subprocess/proc.go:22:1: exported method Process.Pid should have comment or be unexported
shared/subprocess/proc.go:25:1: exported method Process.Stop should have comment or be unexported
shared/subprocess/proc.go:39:1: exported method Process.Start should have comment or be unexported
shared/subprocess/proc.go:74:1: exported method Process.Restart should have comment or be unexported
shared/subprocess/proc.go:85:1: exported method Process.Reload should have comment or be unexported
shared/subprocess/proc.go:99:1: exported method Process.Save should have comment or be unexported
shared/subprocess/proc.go:110:1: exported method Process.Signal should have comment or be unexported
shared/subprocess/proc.go:124:1: exported method Process.Wait should have comment or be unexported
```

You should also cleanup your commits a tiny bit, the second commit includes the rename which makes things confusing, it'd be better to move the rename into the first commit and only have the second commit be the tests.

You'll also want to add the new package to `test/suites/static_analysis.sh` so `golint` is run automatically on it.

File: shared/subprocess/bgpm_test.go [hunk 1]
```diff
@@ -0,0 +1,154 @@
+package subprocess
+
+import (
+	"io"
+	"os"
+	"testing"
+	"time"
+	"strings"
+)
```
Comment (stgraber) [inline shared/subprocess/bgpm_test.go]: Line break before function definition.

File: shared/subprocess/manager.go [hunk 2]
```diff
@@ -0,0 +1,32 @@
+package subprocess
+
+import (
+	"fmt"
+	"gopkg.in/yaml.v2"
+	"io/ioutil"
```
Comment (stgraber) [inline shared/subprocess/manager.go]: Standard import ordering for LXD would have this look like:
```
import (
    "fmt"
    "io/ioutil"

    "gopkg.in/yaml.v2"
)
```

File: shared/subprocess/proc.go [hunk 3]
```diff
@@ -0,0 +1,136 @@
+package subprocess
+
+import (
+	"fmt"
+	"gopkg.in/yaml.v2"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"strings"
+	"syscall"
```
Comment (stgraber) [inline shared/subprocess/proc.go]: Similar comment to other file, the `gopkg.in/yaml.v2` should be separated from the rest as an external import.

File: shared/subprocess/proc.go [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+	pid    int64
+	Name   string   `yaml:"name"`
+	Args   []string `yaml:"args,flow"`
+	Stdin  string   `yaml:"stdin"`
+	Stdout string   `yaml:"stdout"`
+	Stderr string   `yaml:"stderr"`
+}
+
+func (p *Process) Pid() (int64, error) {
+	return p.pid, nil
```
Comment (stgraber) [inline shared/subprocess/proc.go]: Should probably return an error if not running

File: shared/subprocess/proc.go [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+}
+func (p *Process) Stop() error {
```
Comment (stgraber) [inline shared/subprocess/proc.go]: Line break before function definition.

File: shared/subprocess/proc.go [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+		if err != nil {
+			return fmt.Errorf("Could not kill process: %s", err)
+		}
+		return nil
+	} else if err == syscall.ESRCH { //ESRCH error is if process could not be found
+		return fmt.Errorf("Process is not running. Could not kill process")
+	}
+	return fmt.Errorf("Could not kill process: %s", err)
+}
+func (p *Process) Start() error {
```
Comment (stgraber) [inline shared/subprocess/proc.go]: Line break before function definition.

File: shared/subprocess/proc.go [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+	cmd := exec.Command(p.Name, args)
+
+	if p.Stdout != "" {
+		out, err := os.Create(p.Stdout)
+		if err != nil {
+			return fmt.Errorf("Unable to open stdout file: %s", err)
+		}
+		defer out.Close()
+		cmd.Stdout = out
+	}
```
Comment (stgraber) [inline shared/subprocess/proc.go]: Line break

File: shared/subprocess/proc.go [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+	if p.Stdin != "" {
+		in, err := os.Open(p.Stdin)
+		if err != nil {
+			return fmt.Errorf("Unable to open stdin file: %s", err)
+		}
+		defer in.Close()
+		cmd.Stdin = in
+	}
+	if p.Stderr != "" {
```
Comment (stgraber) [inline shared/subprocess/proc.go]: Line break

File: shared/subprocess/proc.go [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+		sterr, err := os.Create(p.Stderr)
+		if err != nil {
+			return fmt.Errorf("Unable to open stderr file: %s", err)
+		}
+		defer sterr.Close()
+		cmd.Stderr = sterr
+	}
+	err := cmd.Start()
```
Comment (stgraber) [inline shared/subprocess/proc.go]: Line break

File: shared/subprocess/proc.go [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+	if err != nil {
+		return fmt.Errorf("Unable to start process: %s", err)
+	}
+	p.pid = int64(cmd.Process.Pid)
```
Comment (stgraber) [inline shared/subprocess/proc.go]: Line break

File: shared/subprocess/proc.go [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+	return nil
+}
+func (p *Process) Restart() error {
```
Comment (stgraber) [inline shared/subprocess/proc.go]: Line break before function definition.

File: shared/subprocess/proc.go [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+	err := p.Stop()
+	if err != nil {
+		return fmt.Errorf("Unable to stop process: %s", err)
+	}
```
Comment (stgraber) [inline shared/subprocess/proc.go]: Line break

File: shared/subprocess/proc.go [hunk 13]
```diff
@@ -0,0 +1,136 @@
...
+	err = p.Start()
+	if err != nil {
+		return fmt.Errorf("Unable to start process: %s", err)
+	}
+	return nil
```
Comment (stgraber) [inline shared/subprocess/proc.go]: Line break

File: shared/subprocess/proc.go [hunk 14]
```diff
@@ -0,0 +1,136 @@
...
+}
+func (p *Process) Reload() error {
```
Comment (stgraber) [inline shared/subprocess/proc.go]: Line break before function definition.

File: shared/subprocess/proc.go [hunk 15]
```diff
@@ -0,0 +1,136 @@
...
+	if err == nil {
+		err = pr.Signal(syscall.SIGHUP)
+		if err != nil {
+			return fmt.Errorf("Could not reload process: %s", err)
+		}
+		return nil
+	} else if err == syscall.ESRCH { //ESRCH error is if process could not be found
+		return fmt.Errorf("Process is not running. Could not reload process")
+	}
+	return fmt.Errorf("Could not reload process: %s", err)
```
Comment (stgraber) [inline shared/subprocess/proc.go]: Line break

File: shared/subprocess/proc.go [hunk 16]
```diff
@@ -0,0 +1,136 @@
...
+	return nil
+}
+func (p *Process) Reload() error {
+	pr, _ := os.FindProcess(int(p.pid))
+	err := pr.Signal(syscall.Signal(0))
+	if err == nil {
+		err = pr.Signal(syscall.SIGHUP)
+		if err != nil {
+			return fmt.Errorf("Could not reload process: %s", err)
+		}
```
Comment (stgraber) [inline shared/subprocess/proc.go]: Line break

File: shared/subprocess/proc.go [hunk 17]
```diff
@@ -0,0 +1,136 @@
...
+		return nil
+	} else if err == syscall.ESRCH { //ESRCH error is if process could not be found
+		return fmt.Errorf("Process is not running. Could not reload process")
+	}
+	return fmt.Errorf("Could not reload process: %s", err)
+}
+func (p *Process) Save(path string) error {
```
Comment (stgraber) [inline shared/subprocess/proc.go]: Line break before function definition.

File: shared/subprocess/proc.go [hunk 18]
```diff
@@ -0,0 +1,136 @@
...
+	dat, err := yaml.Marshal(p)
+	if err != nil {
+		return fmt.Errorf("Unable to serialize process struct to YAML: %s", err)
+	}
+	err = ioutil.WriteFile(path, dat, 0644)
```
Comment (stgraber) [inline shared/subprocess/proc.go]: Line break

File: shared/subprocess/proc.go [hunk 19]
```diff
@@ -0,0 +1,136 @@
...
+	if err != nil {
+		return fmt.Errorf("Unable to write to file %s: %s", path, err)
+	}
+	return nil
```
Comment (stgraber) [inline shared/subprocess/proc.go]: Line break

File: shared/subprocess/proc.go [hunk 20]
```diff
@@ -0,0 +1,136 @@
...
+}
+func (p *Process) Signal(signal int64) error {
```
Comment (stgraber) [inline shared/subprocess/proc.go]: Line break before function definition.

File: shared/subprocess/proc.go [hunk 21]
```diff
@@ -0,0 +1,136 @@
...
+	if err == nil {
+		err = pr.Signal(syscall.Signal(signal))
+		if err != nil {
+			return fmt.Errorf("Could not signal process: %s", err)
+		}
+		return nil
+	} else if err == syscall.ESRCH { //ESRCH error is if process could not be found
+		return fmt.Errorf("Process is not running. Could not signal process")
+	}
+	return fmt.Errorf("Could not signal process: %s", err)
```
Comment (stgraber) [inline shared/subprocess/proc.go]: Line break

File: shared/subprocess/proc.go [hunk 22]
```diff
@@ -0,0 +1,136 @@
...
+}
+func (p *Process) Signal(signal int64) error {
+	pr, _ := os.FindProcess(int(p.pid))
+	err := pr.Signal(syscall.Signal(0))
+	if err == nil {
+		err = pr.Signal(syscall.Signal(signal))
+		if err != nil {
+			return fmt.Errorf("Could not signal process: %s", err)
+		}
+		return nil
```
Comment (stgraber) [inline shared/subprocess/proc.go]: Line break

File: shared/subprocess/proc.go [hunk 23]
```diff
@@ -0,0 +1,136 @@
...
+	} else if err == syscall.ESRCH { //ESRCH error is if process could not be found
+		return fmt.Errorf("Process is not running. Could not signal process")
+	}
+	return fmt.Errorf("Could not signal process: %s", err)
+}
+func (p *Process) Wait() (int64, error) {
```
Comment (stgraber) [inline shared/subprocess/proc.go]: Line break before function definition.

File: shared/subprocess/proc.go [hunk 24]
```diff
@@ -0,0 +1,136 @@
...
+	pr, _ := os.FindProcess(int(p.pid))
+	err := pr.Signal(syscall.Signal(0))
+	if err == nil {
+		procstate, err := pr.Wait()
+		if err != nil {
+			return -1, fmt.Errorf("Could not wait on process: %s", err)
+		}
+		exitcode := int64(procstate.Sys().(syscall.WaitStatus).ExitStatus())
```
Comment (stgraber) [inline shared/subprocess/proc.go]: Line break

File: shared/subprocess/proc.go [hunk 25]
```diff
@@ -0,0 +1,136 @@
...
+		return exitcode, nil
+	}
+	return 0, fmt.Errorf("Process is not running. Could not wait")
```
Comment (stgraber) [inline shared/subprocess/proc.go]: Line break

File: shared/subprocess/testscript/exit1.py [hunk 26]
```diff
@@ -0,0 +1,6 @@
+from time import sleep
+
+print('hello again')
+print('waiting now')
+sleep(4)
+exit(1)
```
Comment (stgraber) [inline shared/subprocess/testscript/exit1.py]: Missing final carriage return

File: shared/subprocess/testscript/signal.py [hunk 27]
```diff
@@ -0,0 +1,15 @@
...
+    return
+
+
+
+signal.signal(signal.SIGHUP, handler)
+signal.signal(10, handler)
+
+signal.pause()
+signal.pause()
+exit(1)
```
Comment (stgraber) [inline shared/subprocess/testscript/signal.py]: Missing final carriage return

File: shared/subprocess/proc.go [hunk 28]
```diff
@@ -0,0 +1,136 @@
...
+	if err == nil {
+		err = pr.Kill()
+		if err != nil {
+			return fmt.Errorf("Could not kill process: %s", err)
+		}
+		return nil
+	} else if err == syscall.ESRCH { //ESRCH error is if process could not be found
+		return fmt.Errorf("Process is not running. Could not kill process")
+	}
+	return fmt.Errorf("Could not kill process: %s", err)
```
Comment (stgraber) [inline shared/subprocess/proc.go]: Line break

File: shared/subprocess/proc.go [hunk 29]
```diff
@@ -0,0 +1,136 @@
...
+}
+func (p *Process) Stop() error {
+	pr, _ := os.FindProcess(int(p.pid))
+	err := pr.Signal(syscall.Signal(0))
+	if err == nil {
+		err = pr.Kill()
+		if err != nil {
+			return fmt.Errorf("Could not kill process: %s", err)
+		}
+		return nil
```
Comment (stgraber) [inline shared/subprocess/proc.go]: Line break

File: shared/subprocess/manager.go [hunk 30]
```diff
@@ -0,0 +1,32 @@
...
+	if err != nil {
+		return nil, fmt.Errorf("Unable to read file %s: %s", path, err)
+	}
+	proc := Process{}
+
+	err = yaml.Unmarshal(dat, &proc)
+	if err != nil {
+		return nil, fmt.Errorf("Unable to parse Process YAML: %s", err)
+	}
+	return &proc, nil
```
Comment (stgraber) [inline shared/subprocess/manager.go]: Line break

File: shared/subprocess/bgpm_test.go [hunk 31]
```diff
@@ -0,0 +1,154 @@
...
+		// Break if finally arrived at end of file
+		if err == io.EOF {
+			break
+		}
+		// Break if error occured
+		if err != nil && err != io.EOF {
+			t.Error("Error reading file: ", err)
+		}
+	}
+	if string(text) != exp {
```
Comment (stgraber) [inline shared/subprocess/bgpm_test.go]: Line break

File: shared/subprocess/bgpm_test.go [hunk 32]
```diff
@@ -0,0 +1,154 @@
...
+		t.Errorf("File output mismatch Expected %s got %s", "hello again\nwaiting now\n", string(text))
+	}
+	// Cleanup
```
Comment (stgraber) [inline shared/subprocess/bgpm_test.go]: Line break

File: shared/subprocess/bgpm_test.go [hunk 33]
```diff
@@ -0,0 +1,154 @@
...
+	if err != nil {
+		t.Error("Failed to get exit code: ", err)
+	} else if ecode != 1 {
+		t.Error("Exit code is not 1: ", ecode)
+	}
+	file, err = os.OpenFile("testscript/out.txt", os.O_RDWR, 0644)
+	if err != nil {
+		t.Error("Could not open file: ", err)
+	}
+	defer file.Close()
```
Comment (stgraber) [inline shared/subprocess/bgpm_test.go]: Line break after the defer

File: shared/subprocess/bgpm_test.go [hunk 34]
```diff
@@ -0,0 +1,154 @@
...
+	err = p.Start()
+	if err != nil {
+		t.Error("Failed to start process: ", err)
+	}
+	ecode, err := p.Wait()
+	if err != nil {
+		t.Error("Failed to get exit code: ", err)
+	} else if ecode != 1 {
+		t.Error("Exit code is not 1: ", ecode)
+	}
```
Comment (stgraber) [inline shared/subprocess/bgpm_test.go]: Line break

File: shared/subprocess/bgpm_test.go [hunk 35]
```diff
@@ -0,0 +1,154 @@
...
+	var text []byte
+	a = append(a, "testscript/exit1.py")
+	p, err := NewProcess("python", a, "", "testscript/out.txt", "")
+	if err != nil {
+		t.Error("Failed process creation: ", err)
+	}
+	err = p.Start()
+	if err != nil {
+		t.Error("Failed to start process: ", err)
+	}
```
Comment (stgraber) [inline shared/subprocess/bgpm_test.go]: Line break

File: shared/subprocess/bgpm_test.go [hunk 36]
```diff
@@ -0,0 +1,154 @@
...
+func TestProcessStartWaitExit(t *testing.T) {
+	var a []string
+	var file *os.File
+	var exp string
+	var text []byte
+	a = append(a, "testscript/exit1.py")
+	p, err := NewProcess("python", a, "", "testscript/out.txt", "")
+	if err != nil {
+		t.Error("Failed process creation: ", err)
+	}
```
Comment (stgraber) [inline shared/subprocess/bgpm_test.go]: Line break

File: shared/subprocess/bgpm_test.go [hunk 37]
```diff
@@ -0,0 +1,154 @@
...
+	err = p.Restart()
+	if err != nil {
+		t.Error("Failed to restart process: ", err)
+	}
+	exitcode, err := p.Wait()
+	if err != nil {
+		t.Error("Could not wait for process: ", err)
+	} else if exitcode != 0 {
+		t.Error("Exit code expected to be 0")
+	}
```
Comment (stgraber) [inline shared/subprocess/bgpm_test.go]: Line break

File: shared/subprocess/bgpm_test.go [hunk 38]
```diff
@@ -0,0 +1,154 @@
...
+		t.Error("Failed to import process: ", err)
+	}
+	err = p.Start()
+	if err != nil {
+		t.Error("Failed to start process: ", err)
+	}
+	err = p.Restart()
+	if err != nil {
+		t.Error("Failed to restart process: ", err)
+	}
```
Comment (stgraber) [inline shared/subprocess/bgpm_test.go]: Line break

File: shared/subprocess/bgpm_test.go [hunk 39]
```diff
@@ -0,0 +1,154 @@
...
+		t.Error("Failed to save process: ", err)
+	}
+	p, err = ImportProcess("testscript/test2.yaml")
+	if err != nil {
+		t.Error("Failed to import process: ", err)
+	}
+	err = p.Start()
+	if err != nil {
+		t.Error("Failed to start process: ", err)
+	}
```
Comment (stgraber) [inline shared/subprocess/bgpm_test.go]: Line break

File: shared/subprocess/bgpm_test.go [hunk 40]
```diff
@@ -0,0 +1,154 @@
...
+		t.Error("Failed to stop process: ", err)
+	}
+	err = p.Save("testscript/test2.yaml")
+	if err != nil {
+		t.Error("Failed to save process: ", err)
+	}
+	p, err = ImportProcess("testscript/test2.yaml")
+	if err != nil {
+		t.Error("Failed to import process: ", err)
+	}
```
Comment (stgraber) [inline shared/subprocess/bgpm_test.go]: Line break

File: shared/subprocess/bgpm_test.go [hunk 41]
```diff
@@ -0,0 +1,154 @@
...
+		t.Error("Failed to start process: ", err)
+	}
+	err = p.Stop()
+	if err != nil {
+		t.Error("Failed to stop process: ", err)
+	}
+	err = p.Save("testscript/test2.yaml")
+	if err != nil {
+		t.Error("Failed to save process: ", err)
+	}
```
Comment (stgraber) [inline shared/subprocess/bgpm_test.go]: Line break

File: shared/subprocess/bgpm_test.go [hunk 42]
```diff
@@ -0,0 +1,154 @@
...
+		t.Error("Failed process creation: ", err)
+	}
+	err = p.Start()
+	if err != nil {
+		t.Error("Failed to start process: ", err)
+	}
+	err = p.Stop()
+	if err != nil {
+		t.Error("Failed to stop process: ", err)
+	}
```
Comment (stgraber) [inline shared/subprocess/bgpm_test.go]: Line break

File: shared/subprocess/bgpm_test.go [hunk 43]
```diff
@@ -0,0 +1,154 @@
...
+	var a []string
+	a = append(a, "testscript/stoprestart.py")
+	p, err := NewProcess("python", a, "", "", "")
+	if err != nil {
+		t.Error("Failed process creation: ", err)
+	}
+	err = p.Start()
+	if err != nil {
+		t.Error("Failed to start process: ", err)
+	}
```
Comment (stgraber) [inline shared/subprocess/bgpm_test.go]: Line break

File: shared/subprocess/bgpm_test.go [hunk 44]
```diff
@@ -0,0 +1,154 @@
...
+}
+
+//tests newprocess, start, stop, save, import, restart, wait
+func TestStopRestart(t *testing.T) {
+	var a []string
+	a = append(a, "testscript/stoprestart.py")
+	p, err := NewProcess("python", a, "", "", "")
+	if err != nil {
+		t.Error("Failed process creation: ", err)
+	}
```
Comment (stgraber) [inline shared/subprocess/bgpm_test.go]: Line break

File: shared/subprocess/bgpm_test.go [hunk 45]
```diff
@@ -0,0 +1,154 @@
...
+            t.Error("Error in reading file ", err)
+        }
+    }
+    if !strings.Contains(string(text), "Called with signal 1"){
+		t.Errorf("Reload failed. File output mismatch. Got %s", string(text))
+	}
+
+	if !strings.Contains(string(text), "Called with signal 10"){
+		t.Errorf("Signal failed. File output mismatch. Got %s", string(text))
+	}
```
Comment (stgraber) [inline shared/subprocess/bgpm_test.go]: Line break

File: shared/subprocess/bgpm_test.go [hunk 46]
```diff
@@ -0,0 +1,154 @@
...
+		_, err = file.Read(text)
+		// Break if finally arrived at end of file
+		if err == io.EOF {
+            break
+        }
+        // Break if error occured
+        if err != nil && err != io.EOF {
+            t.Error("Error in reading file ", err)
+        }
+    }
```
Comment (stgraber) [inline shared/subprocess/bgpm_test.go]: Line break and weird alignment going on here, gofmt should fix that.

File: shared/subprocess/bgpm_test.go [hunk 47]
```diff
@@ -0,0 +1,154 @@
...
+	}
+	defer file.Close()
+
+	var text = make([]byte, 1024)
+	for {
+		_, err = file.Read(text)
+		// Break if finally arrived at end of file
+		if err == io.EOF {
+            break
+        }
```
Comment (stgraber) [inline shared/subprocess/bgpm_test.go]: Line break

File: shared/subprocess/bgpm_test.go (hunk 46 above)
Comment (stgraber) [inline shared/subprocess/bgpm_test.go]: Alignment here is odd

File: shared/subprocess/bgpm_test.go [hunk 48]
```diff
@@ -0,0 +1,154 @@
...
+	time.Sleep(2 * time.Second)
+	p.Reload()
+	time.Sleep(2 * time.Second)
+	p.Signal(10)
+	ecode, err := p.Wait()
+	if err != nil {
+		t.Error("Failed to get exit code ", err)
+	} else if ecode != 1 {
+		t.Error("Exit code is not 1: ", ecode)
+	}
```
Comment (stgraber) [inline shared/subprocess/bgpm_test.go]: Line break

File: shared/subprocess/bgpm_test.go [hunk 49]
```diff
@@ -0,0 +1,154 @@
...
+		t.Error("Failed process creation: ", err)
+	}
+	err = p.Start()
+	if err != nil {
+		t.Error("Failed to start process ", err)
+	}
+	time.Sleep(2 * time.Second)
+	p.Reload()
+	time.Sleep(2 * time.Second)
+	p.Signal(10)
```
Comment (stgraber) [inline shared/subprocess/bgpm_test.go]: Line break

File: shared/subprocess/bgpm_test.go [hunk 50]
```diff
@@ -0,0 +1,154 @@
...
+	a = append(a, "testscript/signal.py")
+	var file *os.File
+	p, err := NewProcess("python", a, "", "testscript/signal_out.txt", "")
+	if err != nil {
+		t.Error("Failed process creation: ", err)
+	}
+	err = p.Start()
+	if err != nil {
+		t.Error("Failed to start process ", err)
+	}
```
Comment (stgraber) [inline shared/subprocess/bgpm_test.go]: Line break

File: shared/subprocess/bgpm_test.go [hunk 51]
```diff
@@ -0,0 +1,154 @@
...
+	"strings"
+)
+func TestSignalHandling(t *testing.T) {
+	var a []string
+	a = append(a, "testscript/signal.py")
+	var file *os.File
+	p, err := NewProcess("python", a, "", "testscript/signal_out.txt", "")
+	if err != nil {
+		t.Error("Failed process creation: ", err)
+	}
```
Comment (stgraber) [inline shared/subprocess/bgpm_test.go]: Line break

File: shared/subprocess/bgpm_test.go [hunk 52]
```diff
@@ -0,0 +1,154 @@
...
+	"os"
+	"testing"
+	"time"
+	"strings"
+)
+func TestSignalHandling(t *testing.T) {
+	var a []string
+	a = append(a, "testscript/signal.py")
+	var file *os.File
+	p, err := NewProcess("python", a, "", "testscript/signal_out.txt", "")
```
Comment (stgraber) [inline shared/subprocess/bgpm_test.go]: Line break

**Comment by stgraber** [comment]
Thanks for this!

I just did first quick pass of static analysis and code styling checks.

**Comment by ribsthakkar** [comment]
Reformatted the code per your comments as well as moving around the rename to the first commit

**Comment by stgraber** [comment]
jenkins: test this please

**Comment by stgraber** [comment]
There's a bit of work needed to get golint to be happy with this still:
```
shared/subprocess/manager.go:10:1: exported function NewProcess should have comment or be unexported
shared/subprocess/manager.go:22:1: exported function ImportProcess should have comment or be unexported
shared/subprocess/proc.go:14:6: exported type Process should have comment or be unexported
shared/subprocess/proc.go:23:1: exported method Process.Pid should have comment or be unexported
shared/subprocess/proc.go:33:1: exported method Process.Stop should have comment or be unexported
shared/subprocess/proc.go:50:1: exported method Process.Start should have comment or be unexported
shared/subprocess/proc.go:90:1: exported method Process.Restart should have comment or be unexported
shared/subprocess/proc.go:104:1: exported method Process.Reload should have comment or be unexported
shared/subprocess/proc.go:119:1: exported method Process.Save should have comment or be unexported
shared/subprocess/proc.go:133:1: exported method Process.Signal should have comment or be unexported
shared/subprocess/proc.go:149:1: exported method Process.Wait should have comment or be unexported
```

**Comment by abbykrish** [comment]
Hi Stephane, I cleared up those golint comments

File: shared/subprocess/bgpm_test.go [hunk 53]
```diff
@@ -0,0 +1,183 @@
...
+		// Break if finally arrived at end of file
+		if err == io.EOF {
+			break
+		}
+
+		// Break if error occured
+		if err != nil && err != io.EOF {
+			t.Error("Error in reading file ", err)
+		}
+	}
```
Comment (stgraber) [inline shared/subprocess/bgpm_test.go]: Line break

File: shared/subprocess/bgpm_test.go [hunk 54]
```diff
@@ -0,0 +1,183 @@
...
+	err = os.Remove("testscript/signal_out.txt")
+	if err != nil {
+		t.Error("Could not delete file ", err)
+	}
+}
+
+//tests newprocess, start, stop, save, import, restart, wait
+func TestStopRestart(t *testing.T) {
+	var a []string
+	a = append(a, "testscript/stoprestart.py")
```
Comment (stgraber) [inline shared/subprocess/bgpm_test.go]: Line break

**Comment by stgraber** [comment]
So other than those two nits, it'd be good if you could:
 - Combine the last commit with the first
 - Replace the python based tests with POSIX shell tests (as we can't guarantee that python will be present on the systems the test run on).

**Comment by ribsthakkar** [comment]
Converted the tests to shell as well!

**Comment by stgraber** [comment]
jenkins: test this please

**Comment by stgraber** [comment]
```
shared/subprocess/bgpm_test.go:59:20: "occured" is a misspelling of "occurred"
shared/subprocess/bgpm_test.go:170:20: "occured" is a misspelling of "occurred"
```

**Comment by ribsthakkar** [comment]
Fixed spelling

**Comment by stgraber** [comment]
jenkins: test this please

**Comment by stgraber** [comment]
@ribsthakkar you now have `go fmt` complaining. Can you run `go fmt ./...` and include those changes in your commits?

**Comment by ribsthakkar** [comment]
Any other formatting things I should run?

**Comment by stgraber** [comment]
jenkins: test this please

**Comment by stgraber** [comment]
We should run `go fmt` (or `gofmt`), `deadcode`, `go vet` and `golint` for the common ones.
We have a couple more special ones that run to detect weirder cases, you can find the whole set in `test/suites/static_analysis.sh`.

**Comment by stgraber** [comment]
Failures are unrelated, so this should be good to go, will do a last review and merge.

**Pull request merged**